app/models/concerns/team_duplication.rb

```diff
@@ -7,215 +7,109 @@ module TeamDuplication
   included do
     attr_accessor :mapping, :original_team, :copy_team
 
-    def self.duplicate(t, user = nil)
-      @mapping = {}
-      @original_team = t
-      @cloned_versions = []
-      begin
-        ActiveRecord::Base.transaction do
-          Version.skip_callback(:create, :after, :increment_project_association_annotations_count)
-          team = t.deep_clone include: [
-            :sources,
-            { projects: [:project_media_projects] },
-            { project_medias: [versions: { if: lambda{|v| v.associated_id.blank? }}]},
-            :team_users,
-            :contacts,
-            :team_tasks
-          ] do |original, copy|
-            @cloned_versions << copy if original.is_a?(Version)
-            self.set_mapping(original, copy) unless original.is_a?(Version)
-            self.copy_image(original, copy)
-            self.versions_log_mapping(original, copy)
-            self.flag_project_medias(original, copy)
-          end
-          team.slug = team.generate_copy_slug
-          team.is_being_copied = true
-          team.save(validate: false)
-          @copy_team = team
-          self.copy_annotations
-          self.update_project_medias
-          self.update_project_media_projects
-          self.copy_relationships(@mapping[:"ProjectMedia"])
-          self.copy_versions(@mapping[:"Version"])
-          self.update_cloned_versions(@cloned_versions)
-          self.create_copy_version(@mapping[:ProjectMedia], user)
-          Version.set_callback(:create, :after, :increment_project_association_annotations_count)
-          team
+    def self.duplicate(t, custom_slug=nil, custom_name=nil)
+      @clones = []
+      @project_id_map = {}
+      @team_id = nil
+      @custom_name = custom_name
+      @custom_slug = custom_slug
+      ActiveRecord::Base.transaction do
+        Version.skip_callback(:create, :after, :increment_project_association_annotations_count)
+        team = t.deep_clone include: [
+          :projects,
+          :contacts,
+          :tag_texts,
+          :team_tasks
+        ] do |original, copy|
+          next if original.is_a?(Version)
+          @clones << {original: original, clone: copy}
+          self.alter_copy_by_type(original, copy)
         end
-      rescue StandardError => e
-        self.log_error(e, t)
-        nil
+        processed_user_ids = self.process_team_bot_installations(t, team)
+        self.process_team_users(t, team, processed_user_ids)
+        team = self.update_team_rules(team)
+        Team.current = team
+        team.save(validate: false)
```

Comment:
Please replace this by `team.save!` - we want to raise an exception if the new team is not valid (for example, slug exists or contains invalid characters).